File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/UnboundModeDemo/UnboundModeDemoMain.pas
unit UnboundModeDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, cxControls, cxLookAndFeels, ActnList, ImgList, Menus, ComCtrls,
  StdCtrls, DemoBasicMain, cxContainer, cxEdit, cxTextEdit, cxStyles, cxTL,
  cxInplaceContainer, cxEditRepositoryItems, cxGraphics, cxCustomData;

type
  TUnboundModeDemoMainForm = class(TDemoBasicMainForm)
    tlPlanets: TcxTreeList;
    clName: TcxTreeListColumn;
    clOrbitNumb: TcxTreeListColumn;
    clOrbits: TcxTreeListColumn;
    clDistance: TcxTreeListColumn;
    clPeriod: TcxTreeListColumn;
    clDiscoverer: TcxTreeListColumn;
    clDate: TcxTreeListColumn;
    clRadius: TcxTreeListColumn;
    clImageIndex: TcxTreeListColumn;
    ilPlanets: TImageList;
    edrepMain: TcxEditRepository;
    edrepCenterText: TcxEditRepositoryTextItem;
    edrepRightText: TcxEditRepositoryTextItem;
    miSeparator1: TMenuItem;
    miDropNodeIndicator: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure tlPlanetsDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure miDropNodeIndicatorClick(Sender: TObject);
  private
    procedure CustomizeColumns;
    procedure LoadData;
  end;

var
  UnboundModeDemoMainForm: TUnboundModeDemoMainForm;

implementation


uses 
  UnboundModeDemoData, ShellAPI, cxDataStorage, cxFormats, SkinDemoUtils;

{$R *.dfm}

procedure TUnboundModeDemoMainForm.CustomizeColumns;
const
  cDistance = 3;
  cPeriod = 4;
  cRadius = 7;
  cImageIndex = 8;
var
  I: Integer;
begin
  with tlPlanets do
  for I := 0 to ColumnCount - 1 do
    if I in [cDistance, cRadius, cImageIndex] then
      Columns[I].DataBinding.ValueTypeClass := TcxIntegerValueType
    else
      if I in [cPeriod] then
      Columns[I].DataBinding.ValueTypeClass := TcxFloatValueType
      else
       Columns[I].DataBinding.ValueTypeClass := TcxStringValueType;
end;

procedure TUnboundModeDemoMainForm.LoadData;
const
  AFileName = 'nineplanets.txt';
  AHeaderLineCount = 2;
  AParentKeyField = 2;
  AKeyField = 0;
  AImageField = 8;

var
  ARecords, AValues: TStringList;
  I: Integer;

  function AddNode(AParentNode: TcxTreeListNode;
    const ARecord: string): TcxTreeListNode;
  var
    S1: string;
    J: Integer;
    V: Variant;
  begin
    Result := AParentNode.AddChild;
    AValues.CommaText := ARecord;
    for J := 0 to AValues.Count - 1 do
      if AValues.Strings[J] <> '-' then
      begin
        S1 := AValues.Strings[J];
        if Pos('.', S1) <> 0 then
          S1[Pos('.', S1)] := dxFormatSettings.DecimalSeparator;
        V := S1;
        if not VarIsNull(V) then
          Result.Values[J] := V;
      end;
    Result.ImageIndex :=  Result.Values[AImageField];
    Result.SelectedIndex := Result.Values[AImageField];
  end;

  procedure AddNodes(AParentNode: TcxTreeListNode;
     const AParentKeyValue: string);
     function GetFieldValue(ARecord: string; AFieldIndex: Integer): string;
     begin
       AValues.CommaText := ARecord;
       Result := AValues.Strings[AFieldIndex];
     end;
  var
    J: Integer;
    ANode: TcxTreeListNode;
  begin
    for J := 0 to ARecords.Count - 1 do
      if GetFieldValue(ARecords.Strings[J], AParentKeyField) = AParentKeyValue then
      begin
        ANode := AddNode(AParentNode, ARecords.Strings[J]);
        AddNodes(ANode, GetFieldValue(ARecords.Strings[J], AKeyField));
      end;
  end;

begin
  if not FileExists(AFileName) then
    raise Exception.Create('Data file not found');

  ARecords := TStringList.Create;
  AValues := TStringList.Create;

  tlPlanets.BeginUpdate;
  with ARecords do
    try
      LoadFromFile(AFileName);
      for I := 0 to AHeaderLineCount - 1 do
        Delete(0);
      AddNodes(tlPlanets.Root, '-');
    finally
      tlPlanets.EndUpdate;
      ARecords.Free;
      AValues.Free;
    end;
end;

procedure TUnboundModeDemoMainForm.FormCreate(Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code}
  inherited;
  CustomizeColumns;
  LoadData;
  tlPlanets.FullCollapse;
  tlPlanets.Root[0].Expanded := True;

//}
end;

procedure TUnboundModeDemoMainForm.FormShow(Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code

  ShowMessage('WARNING: tutorial not completed. First, please apply the steps '+
              'shown in the doc file');

//}
end;

procedure TUnboundModeDemoMainForm.tlPlanetsDragOver(Sender,
  Source: TObject; X, Y: Integer; State: TDragState; var Accept: Boolean);
begin
  inherited;
  //
end;

procedure TUnboundModeDemoMainForm.miDropNodeIndicatorClick(
  Sender: TObject);
begin
  tlPlanets.OptionsView.DropNodeIndicator := GetMenuItemChecked(Sender);
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/UnboundModeDemo/UnboundModeDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "UnboundModeDemoMain.h"
#include "UnboundModeDemoData.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxEditRepositoryItems"
#pragma link "cxInplaceContainer"
#pragma link "cxTextEdit"
#pragma link "cxTL"
#pragma resource "*.dfm"
TUnboundModeDemoMainForm *UnboundModeDemoMainForm;

  const int HeaderLineCount = 2;
  const int ParentKeyField = 2;
  const int KeyField = 0;
  const int ImageField = 8;
  const int DistanceColIndex = 3;
  const int PeriodColIndex = 4;
  const int RadiusColIndex = 7;
  const int ImageIndexColIndex = 8;

//---------------------------------------------------------------------------
__fastcall TUnboundModeDemoMainForm::TUnboundModeDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TUnboundModeDemoMainForm::FormShow(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code *

  ShowMessage("WARNING: tutorial not completed. First, please apply the steps "
    "shown in the doc file");

//*/
}
//---------------------------------------------------------------------------

void __fastcall TUnboundModeDemoMainForm::FormCreate(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code */
  TDemoBasicMainForm::FormCreate(Sender);
  CustomizeColumns();
  LoadData();
  tlPlanets->FullCollapse();
  tlPlanets->Root->Items[0]->Expanded = true;

//*/
}
//---------------------------------------------------------------------------
void __fastcall TUnboundModeDemoMainForm::miDropNodeIndicatorClick(TObject *Sender)
{
  tlPlanets->OptionsView->DropNodeIndicator = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TUnboundModeDemoMainForm::CustomizeColumns()
{
  const String DecimalSeparator = '.';
  for (int I = 0; I < tlPlanets->ColumnCount; I++)
    if (I == DistanceColIndex || I == RadiusColIndex || I == ImageIndexColIndex)
      tlPlanets->Columns[I]->DataBinding->ValueTypeClass =
        (TcxValueTypeClass)__classid(TcxIntegerValueType);
    else{
      if (I == PeriodColIndex)
        tlPlanets->Columns[I]->DataBinding->ValueTypeClass =
          (TcxValueTypeClass)__classid(TcxFloatValueType);
      else
        tlPlanets->Columns[I]->DataBinding->ValueTypeClass =
          (TcxValueTypeClass)__classid(TcxStringValueType);
    }
}
//---------------------------------------------------------------------------

void __fastcall TUnboundModeDemoMainForm::LoadData()
{
  if (!FileExists(FileName))
    throw Exception("Data file not found");

  FRecords = new TStringList();
  FValues = new TStringList();

  tlPlanets->BeginUpdate();
  __try{
    FRecords->LoadFromFile(FileName);
    for (int I = 0; I < HeaderLineCount; I++)
      FRecords->Delete(0);
    AddNodes(tlPlanets->Root, "-");
  }
  __finally{
    tlPlanets->EndUpdate();
    delete FRecords;
    delete FValues;
  }
}
//---------------------------------------------------------------------------

TcxTreeListNode* __fastcall TUnboundModeDemoMainForm::AddNode(
  TcxTreeListNode *AParentNode, String ARecord)
{
  String S;
  TcxTreeListNode *Result = AParentNode->AddChild();
  FValues->CommaText = ARecord;
  for (int J = 0; J < FValues->Count; J++)
    if (FValues->Strings[J] != "-"){
      S = FValues->Strings[J];
      if (S.Pos('.')!=0) {S[S.Pos('.')] = dxFormatSettings->DecimalSeparator;}
      Variant V = S;
      if (!VarIsNull(V))
        Result->Values[J] = V;
    }
  Result->ImageIndex =  Result->Values[ImageField];
  Result->SelectedIndex = Result->Values[ImageField];
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TUnboundModeDemoMainForm::AddNodes(TcxTreeListNode *AParentNode,
  String AParentKeyValue)
{
  TcxTreeListNode *ANode;
  for (int J = 0; J < FRecords->Count; J++)
    if (GetFieldValue(FRecords->Strings[J], ParentKeyField) == AParentKeyValue){
      ANode = AddNode(AParentNode, FRecords->Strings[J]);
      AddNodes(ANode, GetFieldValue(FRecords->Strings[J], KeyField));
    }
}
//---------------------------------------------------------------------------

String __fastcall TUnboundModeDemoMainForm::GetFieldValue(String ARecord,
  int AFieldIndex)
{
 FValues->CommaText = ARecord;
 return FValues->Strings[AFieldIndex];
}
//---------------------------------------------------------------------------


void __fastcall TUnboundModeDemoMainForm::tlPlanetsDragOver(
      TObject *Sender, TObject *Source, int X, int Y, TDragState State,
      bool &Accept)
{
//        
}
//---------------------------------------------------------------------------

